scripts/e2e/android-devices.js

```diff
@@ -0,0 +1,110 @@
+const { getCommandLineResponse } = require('./command-line-helpers');
+
+const DEFAULT_CI_AVD_NAME = 'e2e_emulator';
+const isRunningCI = !!process.env.CI;
+
+const envVarKeys = /** @type {const} */ ({
+  adbSerial: 'RNS_DEVICE_SERIAL',
+  avdName: 'RNS_AVD_NAME',
+});
+
+const passedAdbSerial = process.env[envVarKeys.adbSerial];
+
+function resolveAttachedAndroidDeviceSerial() {
+  if (passedAdbSerial) return passedAdbSerial;
+  const connectedNonEmulators = getDeviceIds().filter(
+    deviceID => !deviceID.startsWith('emulator'),
+  );
+  if (connectedNonEmulators.length === 0) {
+    throw new Error('No physical devices attached.');
+  } else if (connectedNonEmulators.length === 1) {
+    return connectedNonEmulators[0];
+  } else
+    throw new Error(
+      `Connected devices: ${connectedNonEmulators.join(', ')}\nUnplug ${
+        connectedNonEmulators.length - 1
+      } device(s) or use ${
+        envVarKeys.adbSerial
+      } environment variable to select a specific one.`,
+    );
+}
```

Comment:
nit: mixing throwing exceptions & returning values doesn't look good, I'd propose to reorganize this code a little bit:
```js
if (connectedNonEmulators.length === 0) {
// throw
}

if (connectedNonEmulators.length > 1) {
// throw
}

return ...
```